llvm/lib/Analysis/IR2VecAnalysis.cpp

```diff
@@ -0,0 +1,435 @@
+  }
+  assert(Vocabulary.size() > 0 && "Vocabulary is empty");
+
+  unsigned Dim = Vocabulary.begin()->second.size();
+  assert(Dim > 0 && "Dimension of vocabulary is zero");
+  assert(std::all_of(Vocabulary.begin(), Vocabulary.end(),
+                     [Dim](const std::pair<StringRef, Embedding> &Entry) {
+                       return Entry.second.size() == Dim;
+                     }) &&
+         "All vectors in the vocabulary are not of the same dimension");
+  return Error::success();
+}
+
+IR2VecVocabAnalysis::Result
+IR2VecVocabAnalysis::run(Module &M, ModuleAnalysisManager &AM) {
+  auto Ctx = &M.getContext();
+  if (VocabFile.empty()) {
+    // FIXME: Use default vocabulary
+    Ctx->emitError("IR2Vec vocabulary file path not specified");
+    return IR2VecVocabResult(); // Return invalid result
+  }
+  if (auto Err = readVocabulary()) {
+    handleAllErrors(std::move(Err), [&](const ErrorInfoBase &EI) {
+      Ctx->emitError("Error reading vocabulary: " + EI.message());
+    });
+    return IR2VecVocabResult();
+  }
+  return IR2VecVocabResult(std::move(Vocabulary));
+}
+
+// ==----------------------------------------------------------------------===//
+// IR2VecResult and IR2VecAnalysis
+//===----------------------------------------------------------------------===//
+
+IR2VecResult::IR2VecResult(
+    const SmallMapVector<const Instruction *, Embedding, 128> &&InstMap,
+    const SmallMapVector<const BasicBlock *, Embedding, 16> &&BBMap,
+    const Embedding &&FuncVector)
+    : InstVecMap(std::move(InstMap)), BBVecMap(std::move(BBMap)),
+      FuncVector(std::move(FuncVector)), Valid(true) {}
+
+const SmallMapVector<const Instruction *, Embedding, 128> &
+IR2VecResult::getInstVecMap() const {
+  assert(Valid);
+  return InstVecMap;
+}
+const SmallMapVector<const BasicBlock *, Embedding, 16> &
+IR2VecResult::getBBVecMap() const {
+  assert(Valid);
+  return BBVecMap;
+}
+const Embedding &IR2VecResult::getFunctionVector() const {
+  assert(Valid);
+  return FuncVector;
+}
+
+IR2VecAnalysis::Result IR2VecAnalysis::run(Function &F,
+                                           FunctionAnalysisManager &FAM) {
+  auto *VocabRes = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F)
+                       .getCachedResult<IR2VecVocabAnalysis>(*F.getParent());
+  auto Ctx = &F.getContext();
+  if (!VocabRes->isValid()) {
+    Ctx->emitError("IR2Vec vocabulary is invalid");
+    return IR2VecResult();
+  }
+
+  auto Dim = VocabRes->getDimension();
+  if (Dim <= 0) {
+    Ctx->emitError("IR2Vec vocabulary dimension is zero");
+    return IR2VecResult();
+  }
+
+  auto Vocabulary = VocabRes->getVocabulary();
+  std::unique_ptr<Embeddings> Emb;
+  switch (IR2VecMode) {
+  case IR2VecKind::Symbolic:
+    Emb = std::make_unique<Symbolic>(F, Vocabulary, Dim);
+    break;
+  case IR2VecKind::FlowAware:
+    // FIXME: Add support for flow-aware embeddings
+  default:
+    Ctx->emitError("Invalid IR2Vec mode");
```

Comment:
Maybe move this into a factory function like `Create(IR2VecMode` as part of the Embeddings class and return ErrorOr<unique_ptr>.

Reply:
Changed.